docs/proposals/scheduling/20220629-fine-grained-device-scheduling.md

```diff
@@ -0,0 +1,419 @@
+
+We abstract GPU resources into different dimensions:
+
+- `koordinator.sh/gpu-core` represents the computing capacity of the GPU. Similar to K8s MilliCPU, we abstract the total computing power of GPU into one hundred, and users can apply for the corresponding amount of GPU computing power according to their needs.
+- `koordinator.sh/gpu-memory` represents the memory capacity of the GPU in bytes.
+- `koordinator.sh/gpu-memory-ratio` represents the percentage of the GPU's memory.
+
+Assuming that node A has 4 GPU instances, and the total memory of each instance is 8GB, when device reporter reports GPU capacity information to `Node.Status.Allocatable`, it no longer reports nvidia.com/gpu=4, but reports the following information:
+
+```yaml
+status:
+  capacity:
+    koordinator.sh/gpu-core: 400
+    koordinator.sh/gpu-memory: "32GB"
+    koordinator.sh/gpu-memory-ratio: 400
+  allocatable:
+    koordinator.sh/gpu-core: 400
+    koordinator.sh/gpu-memory: "32GB"
+    koordinator.sh/gpu-memory-ratio: 400
+```
+
+For the convenience of users, an independent resource name `koordinator.sh/gpu` is defined. For example, when a user wants to use half of the computing resources and memory resources of a GPU instance, the user can directly declare `koordinator.sh/gpu: 50`, and the scheduler will convert it to `koordinator.sh/gpu-core: 50, koordinator.sh/gpu-memory-ratio: 50`
+
+For other devices like RDMA and FPGA, the node has 1 RDMA and 1 FGPA, will report the following information:
+
+```yaml
+status:
+  capacity:
+    koordinator.sh/rdma: 100
+    koordinator.sh/fpga: 100
+  allocatable:
+    koordinator.sh/rdma: 100
+    koordinator.sh/fpga: 100
+```
+
+Why do we need `koordinator.sh/gpu-memory-ratio` and `koordinator.sh/gpu-memory` ? 
+When user apply 0.5/0.25 GPU, the user don't know the exact memory total bytes per GPU, only wants to use 
+half or quarter percentage of memory, so user can request the GPU memory with `koordinator.sh/gpu-memory-ratio`. 
+When scheduler assigned Pod on concrete node, scheduler will translate the `koordinator.sh/gpu-memory-ratio` to `koordinator.sh/gpu-memory` by the formulas:  ***allocatedMemory = totalMemoryOf(GPU)  * `koordinator.sh/gpu-memory-ratio`***, so that the GPU isolation can work.
+
+During the scheduling filter phase, the scheduler will do special processing for `koordinator.sh/gpu-memory` and `koordinator.sh/gpu-memory-ratio`. When a Pod specifies `koordinator.sh/gpu-memory-ratio`, the scheduler checks each GPU instance on each node for unallocated or remaining resources to ensure that the remaining memory on each GPU instance meets the ratio requirement.
+
+If the user knows exactly or can roughly estimate the specific memory consumption of the workload, he can apply for GPU memory through `koordinator.sh/gpu-memory`. All details can be seen below.
+
+Besides, when dimension's value > 100, means Pod need multi-devices. now only allow the value can be divided by 100.
```

Comment:
If the value of a container's gpu-core is greater than 100 and cannot be divided by 100 (e.g. 101), will the pod be rejected?

Reply:
yes, should be rejected.

Reply:
is it rejected by the webhook, or by the scheduler?

Reply:
by scheduler in Prefilter step.

Reply:
koordinator.sh/gpu-core should be something like: 25, 50, 75, 100, 200, 300. otherwise it will be rejected by scheduler in Prefilter step.

Reply:
if a pod is rejected by the scheduler, it will go into `Pending` phase, and the scheduler will keep retrying to schedule the pod. I think this retry may be useless, and may increase the load on the scheduler. would it be better to reject the pod in the webhook?

Reply:
would it would be better to reject the pod in the webhook.